src/mechanics/IGA/NURBS.h

```diff
@@ -0,0 +1,540 @@
+                saved = left[j - r] * temp;
+            }
+            ndu(j, j) = saved;
+        }
+
+        Eigen::MatrixXd basisFctsDerivatives(derivativeOrder + 1, degree + 1);
+
+        for (int j = 0; j <= degree; j++)
+            basisFctsDerivatives(0, j) = ndu(j, degree);
+
+        // the shape functions
+        Eigen::MatrixXd a(2, degree + 1);
+        for (int r = 0; r <= degree; r++)
+        {
+            int s1 = 0;
+            int s2 = 1;
+
+            a(0, 0) = 1.0;
+
+            int j = 0;
+            for (int k = 1; k <= derivativeOrder; k++)
+            {
+                double d = 0.;
+                int rk = r - k;
+                int pk = degree - k;
+                if (r >= k)
+                {
+                    a(s2, 0) = a(s1, 0) / ndu(pk + 1, rk);
+                    d = a(s2, 0) * ndu(rk, pk);
+                }
+                int j1 = 0;
+                if (rk >= -1)
+                    j1 = 1;
+                else
+                    j1 = -rk;
+
+                int j2 = 0;
+                if (r - 1 <= pk)
+                    j2 = k - 1;
+                else
+                    j2 = degree - r;
+
+                for (j = j1; j <= j2; j++)
+                {
+                    a(s2, j) = (a(s1, j) - a(s1, j - 1)) / ndu(pk + 1, rk + j);
+                    d += a(s2, j) * ndu(rk + j, pk);
+                }
+
+                if (r <= pk)
+                {
+                    a(s2, k) = -a(s1, k - 1) / ndu(pk + 1, r);
+                    d += a(s2, k) * ndu(r, pk);
+                }
+
+                basisFctsDerivatives(k, r) = d;
+                j = s1;
+                s1 = s2;
+                s2 = j;
+            }
+        }
+
+        int r = degree;
+        for (int k = 1; k <= derivativeOrder; k++)
+        {
+            for (int j = 0; j <= degree; j++)
+                basisFctsDerivatives(k, j) *= r;
+            r *= (degree - k);
+        }
+
+        return basisFctsDerivatives;
+    }
+
+    //! @brief ... calculates the basisfunctions and derivatives, see Piegl/Tiller 'NURBS Book' 2nd ed., Page 72
+    //! @param der ... up to derivative der
+    //! @param parameter ... parameter at which the functions are calculated
+    //! @param spanIdx ... spanIdx to the parameter
+    //! @param degree ... degree of the given NURBS curve
+    //! @param knots ... knot vector
+    //! @param weights ... weights vector
+    //! @return Eigen::VectorXd ... vector containing the shape functions (no derivatives)
+    static Eigen::VectorXd BasisFunctionsAndDerivativesRational(int derivativeOrder, double parameter, int spanIdx,
+                                                                int degree, const std::vector<double>& knots,
+                                                                const std::vector<double>& weights)
+    {
+        assert(derivativeOrder >= 0 || derivativeOrder <= 2);
```

Comment:
derivativeOrder values of -12 or 6174 will not trigger this assertion. You should use &&.

Reply:
Or even `and`.